Bug fix: Refer to unpublished but editable Guide pages

Adnan-cds opened the pull request:
Editors should be able to edit Guide overview pages when those pages refer to one or
more unpublished Guide pages.

Similarly, Editors should be able to refer to an unpublished Guide overview
page from a Guide page.

Adding patch for core bug (?).
@see https://www.drupal.org/project/drupal/issues/2845144

Resolves #36 #40

Adnan-cds committed:
Bug fix: Refer to unpublished but editable Guide pages.

Editors should be able to edit Guide overview pages when they refer to one or
more unpublished Guide pages.

Similarly, Editors should be able to refer to an unpublished Guide overview
page.

Adding patch for core bug (?).

@see https://www.drupal.org/project/drupal/issues/2845144

Adnan-cds commented:
@todo Add a Kernel test.

ekes commented:
How does work / what is the situation with LocalGov Workflows: without the patch, and with the patch?

Adnan-cds commented:
Hi Ekes, thanks for mentioning localgov_workflows.  It didn't occur to me before you mentioned it.  I have just enabled the localgov_workflows module in my local site and it has not made any difference without the patch.  At the same time it doesn't conflict with the patch either.  This is understandable given that the [Entity reference selection plugin of the node module](https://git.drupalcode.org/project/drupal/-/blob/9.3.x/core/modules/node/src/Plugin/EntityReferenceSelection/NodeSelection.php) is fairly basic.  It doesn't take content moderation into account at all.

Sorry, I have just marked this pull request as Draft while I finish the Kernel test.  Hopefully all should be done in an hour or so and then I will mark it as Ready for review again.

ekes commented:
> This is understandable given that the Entity reference selection plugin of the node module is fairly basic. 
```
!count($this->moduleHandler->getImplementations('node_grants'))
```
I'd have assumed workflow would be adding access control via node_grants?

I'd think the main thing for integration with workflow though would be that users can reference content in moderation (that they have access to), on pages that can also be in moderation, so they can get a preview of what it would look like with the content before publish both?
For non-access it should be that they can't add or remove references that they don't have access to, but can edit the rest of the entity?

Adnan-cds commented:
> I'd have assumed workflow would be adding access control via node_grants?

Nah, content_moderation doesn't implement hook_node_grants() at all.

> I'd think the main thing for integration with workflow though would be that users can reference content in moderation (that they have access to), on pages that can also be in moderation, so they can get a preview of what it would look like with the content before publish both?

That would be a really good feature to have.  The patch doesn't address that of course.  I can think of two ways to achieve this:
- Use a custom tailored Entity reference view for Guide page (Or Guide Overview page) selection.
- Write a custom Entity reference selection plugin.

Let me know your thoughts.

Adnan-cds committed:
Bug fix: Refer to unpublished but editable Guide pages.

Kernel test to verify usage of *unpublished* Guide overview page and Guide page
as parent and child respectively.

Adnan-cds commented:
I have finished adding the Kernel test.  So this is ready for review again.  We will of course keep exploring better alternatives.

ekes commented:
I think we should move this up to the distro itself?

Off the top of my head this is an issue that occurs on at least:
 * services landingpage; 
 * step-by-step; 
 * as well as guides.

By applying this patch it will also have an impact on at least those - hopefully positive. 
Some (double-)checks we should do before committing this that I can already think of: 
 * check display of those pages with references to unpublished content; 
 * what happens when editors that _don't_ have the permission to unpublished content edit the pages with references to them.

As the patch also only considers user permissions (access unpublished content) and not if users can access it because of content moderation / workflow I *think* (we should check) it's safe for that; but we should also probably specify a follow-up issue for that brain-melt.

Adnan-cds commented:
> I think we should move this up to the distro itself?

I have no problem with that.  Do you want me to raise a pull request for the "localgov" profile?

> check display of those pages with references to unpublished content;

Service related pages highlight unpublished child pages.  But Guides and Step-by-steps don't.  Dan is [doing something](https://github.com/localgovdrupal/localgov_guides/pull/68) about it thankfully.

> what happens when editors that don't have the permission to unpublished content edit the pages with references to them.

 if I understand the behaviour of the code correctly, editors without access to child Guide pages will continue to see the same error messages that everyone is now seeing in Guide overview pages.

> As the patch also only considers user permissions (access unpublished content) and not if users can access it because of content moderation / workflow I think (we should check) it's safe for that;

Yeah, that would be good.

ekes commented:
> >   I think we should move this up to the distro itself?
> I have no problem with that. Do you want me to raise a pull request for the "localgov" profile?

I've collated what I could find together:-
localgovdrupal/localgov#334

Adnan-cds committed:
Revert "Bug fix: Refer to unpublished but editable Guide pages."

This reverts commit 24e5e5322b38c6646c94e168f28a024c6ccfcdda.

Adnan-cds commented:
I have removed the patch in favour of the [pull request Andy has raised](https://github.com/localgovdrupal/localgov/pull/336).  So this change is now blocked by [localgov:#336](https://github.com/localgovdrupal/localgov/pull/336) @andybroomfield @ekes

Adnan-cds commented:
Closing this as the code has moved to the [localgov profile](https://github.com/localgovdrupal/localgov/pull/336).  Thanks everyone for making this happen.  Looking forward to deploying it.  Our Editors have been eagerly waiting for this fix :)
